### PR TITLE
Jetpack Connect Authorize: XMLRPC error fallback

### PR DESCRIPTION
**Pull request by roccotripaldi**
What's going on:
Currently in Jetpack, the connection process happens in wp-admin. Here in calypso we are attempting to do it via an XMLRPC request. Some hosts limit XMLRP requests, which will result in more failed connections using the calypso flow.

This PR attempts to detect XMLRPC errors and fallback to the flow within wp-admin.

To test:
- get this branch going at calypso.localhost:3000
- go to wp-admin dashboard of one of your unconnected Jetpack sites
- click 'connect'
- on the next screen, add `&jetpack_connect_test=calypso&calypso_env=development` to the (really long) URL in your address bar, and press enter
- disable XMLRPC on your Jetpack site, ie rename xmlrpc.php to something else
- click 'Approve' on the calypso screen
- ensure that you receive this improved error screen

screen shot 2016-08-22 at 4 34 22 pm

cc: @kraftbj @johnHackworth @tyxla 

Test live: https://calypso.live/?branch=update/jetpack-connect-xmlrpc-error-fallback

### CODE REVIEW

**Comment by roccotripaldi** [comment]
@ebinnion - i do not think this should affect the SSO process. XMLRPC errors that occur during SSO will still get the error notice `Error connecting your site. Please contact support` because if we instead complete their connection in wp-admin, the SSO flow will be gone. What are your thoughts?

**Comment by ebinnion** [comment]
For SSO, I think it would be ideal if we just redirected the user back to their site if the authorize fails. cc @rickybanister though

**Comment by roccotripaldi** [comment]
@rickybanister et al, do you like what i've proposed in the screen cast, where we simply and silently redirect folks back to their site, or do you like something like this that shows a bit more feedback?

screen shot 2016-08-17 at 5 17 47 pm

Personally, I'm leaning towards showing some feedback.

**Comment by tyxla** [comment]
I've tested the PR and it worked well for me :+1:.

My two cents: I think a combination of showing feedback and redirecting to the Jetpack site would work well. Since we can use the wp-admin flow to finalize the connection when XMLRPC is disabled, lets use it in that case. But we should let the user know that XMLRPC is disabled, and this would cause issues with Jetpack in the future.

**Comment by rickybanister** [comment]
@roccotripaldi I can see the value of providing feedback, but it might be easier for folks if we redirected more quietly. If they click away from here they will disrupt the process and possibly miss out on some of the onboarding stuff. It might be nice to alert them to the xmlrpc problem after the fact or in some asynchronous way—a notification? Something like "while connecting we noticed an error with xxxx, you can fix this by doing xxxx or contacting your hosting provider". 

That's just a thought. If there's overwhelming support for showing the details, I'm fine with that as well.

**Comment by rickybanister** [comment]
I take back my last comment (after hearing more from @roccotripaldi)

Let's show more detail on the Calypso side, but change the interface to this:

Use a notice-action in the notice
`We had trouble connecting | Try again ->`

Explain what happened in the text below that, but without any links except for contacting support (no need for external link icon, just link in the text). We can work on some text together that's more friendly about why it failed and what will happen if you 'try again'. 

And finally, let's leave the original primary action button, but disable it. That will create continuity for the user. They clicked it, there was a problem, they can't re-click it, but they can use the notice to fix the situation.

**Comment by roccotripaldi** [comment]
@rickybanister - i added your changes. here is an updated screen cast: https://cloudup.com/cN-Y5Fs7vaZ

and screen shot:
screen shot 2016-08-22 at 4 34 22 pm

File: client/state/jetpack-connect/selectors.js [hunk 1]
```diff
@@ -21,4 +21,17 @@ const getFlowType = function( state, site ) {
...
 };
 
-export default { isCalypsoStartedConnection, getFlowType };
+const hasXmlrpcError = function( state ) {
+	return (
+		state.authorizeError &&
+		state.authorizeError.message &&
+		state.authorizationCode &&
+		(
+			state.authorizeError.message.indexOf( 'transport error' ) > -1 ||
```
Comment (tyxla) [inline client/state/jetpack-connect/selectors.js]: Is it necessary to list all types of errors? Maybe just checking for the existence of `error` would work as fine?

**Comment by tyxla** [comment]
I've tested the new PR and it worked fine :+1:.

Several things I've noticed that I think are worth mentioning:

1. I've left a couple of minor inline comments.
2. `hasXmlrpcError()` could use tests - I'd be happy to address this in https://github.com/Automattic/wp-calypso/pull/7404 or in a separate PR.
3. For some reason, when I click the "Connect" button, I'm presented with the old authorization screen. Although this is not directly related with this PR, is this the expected behavior?
3. I'd personally remove the "If that does not work, contact support" in favor of the "Get help connecting your site" button which does the same, and is down there anyway.

**Comment by roccotripaldi** [comment]
Thanks for the review @tyxla 
I agree on all points. I will handle tests in this PR later day.

> For some reason, when I click the "Connect" button, I'm presented with the old authorization screen. Although this is not directly related with this PR, is this the expected behavior?

Yep. Only 10% of users are getting the new Calypso flow.

**Comment by rickybanister** [comment]
@roccotripaldi that looks much better, I just want to improve the text area a bit more. It looks a little unfriendly with the technical error code and the somewhat short sentence about contacting support. The notice looks good to go, however.

Let's try something like:
```
We had trouble connecting. | Try again

WordPress.com was unable to reach your site and approve the connection. Try again by clicking the button above; if that doesn't work you may need to contact support.
```

**Comment by roccotripaldi** [comment]
thanks @rickybanister - i've added that text
@tyxla - i've added some tests!

**Comment by tyxla** [comment]
LGTM! 🚢 

Attaching a screenshot of the updated version for easier design feedback:

**Comment by johnHackworth** [comment]
This looks great, rocco. I think it's ready to go, so I'm going to mark it as shippable. If Rick or someone else wants to have any further improvement in, we can do it in another branch

**Comment by johnHackworth** [comment]
Also not for this PR, but ... would it be doable to provide a link to the old auth screen, instead of only showing the current error?

**Comment by rickybanister** [comment]
@johnHackworth I was under the impression that the 'try again' link would be the link to try again via the old way?

I noticed a couple things—I'm not sure we need the error code anymore since we are hoping to describe the error more plainly in the text above. Also, we should link the 'contact support' string as it was previously, just without the external link icon. Even though there's a help link in the footer, it'd be nice to link up the text as the user reads it.

**Comment by roccotripaldi** [comment]
>would it be doable to provide a link to the old auth screen, instead of only showing the current error

@johnHackworth - yep, the try again button sends them back to wp-admin with a `$_GET` request containing all the info needed to complete the connection much like the old flow does.

> I'm not sure we need the error code anymore since we are hoping to describe the error more plainly in the text above

@rickybanister  I'd like to leave the error. Currently, if the connection fails in the wp-admin flow, the error details are displayed in wp-admin. And that info is helpful for HEs. If for any reason in Calypso we cannot fallback and complete the connection in wp-admin, it'd be nice to have some info for HEs to start with.

**Comment by kraftbj** [comment]
+1 to leaving the error code information in. Even if the old way works, that error information can be helpful to have available. We've been chasing ghost errors with the 10% seeing this flow now and it's nearly impossible for us to debug without any hint what may be problematic with XMLRPC.

**Comment by johnHackworth** [comment]
oops, sorry, I thought the 'try again' button just tried to launch the remote auth process again and I didn't even tested it :D

**Pull request merged**